File: matrixMult.java
 public class matrixMult {

   private static long start_time = 0;
   private static long end_time = 0;
   private static double totalTime = 0;

   public static void main(String[] args) {

      //Matrix1
      double[][] arr1 = new double[][]{
        { 1.0, 2.0, 3.0, 4.0, 5.0 },
        { 6.0, 7.0, 8.0, 9.0, 10.0 },
        { 11.0, 12.0, 13.0, 14.0, 15.0 },
        { 16.0, 17.0, 18.0, 19.0, 20.0 },
        { 21.0, 22.0, 23.0, 24.0, 25.0 }
      };

      //Matrix2
      double[][] arr2 = new double[][]{
        { 1.0, 2.0, 3.0, 4.0, 5.0 },
        { 6.0, 7.0, 8.0, 9.0, 10.0 },
        { 11.0, 12.0, 13.0, 14.0, 15.0 },
        { 16.0, 17.0, 18.0, 19.0, 20.0 },
        { 21.0, 22.0, 23.0, 24.0, 25.0 }
      };

      //Matrix Multiplication
      double[][] arr3 = mult(arr1, arr2);
      System.out.println("---MATRIX MULT RESULT---\n" + toString(arr3));

      //Run-time Analysis
      for (int i=0; i < 10; i++){
         start_time = System.nanoTime();
         double[][] result = mult(arr1, arr2);
         end_time = System.nanoTime();

         totalTime += (end_time - start_time);
      }

      double avgTime = totalTime/10.0;
      System.out.println("Java Matrix Mult Average time taken: " + avgTime + " nanoseconds.");
   }

   /**
   * This is the matrix multiplication method on nxn matrices.
   * @param double a[][], double b[][] Two matrices.
   * @return The product of the two matrices.
   */
   public static double[][] mult(double a[][], double b[][]){
      if(a.length == 0) {
         return new double[0][0];
      }
      if(a[0].length != b.length) {
         return null; //invalid dims
      }
    
      int n = a[0].length;
    
      double result[][] = new double[n][n];
    
      for (int i = 0; i < n; i++) {
         for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
               result[i][j] += a[i][k] * b[k][j];
            }
         }
      }
      return result;
   }


   /**
   * This method turns the matrix into a printable value
   * @param double a[][] A matrix.
   * @return String 
   */
   public static String toString(double a[][]) {
      String matrix = "";

      int n = a[0].length;

      for (int row=0; row < n; row++) {
         String matrixRow = "";
         for (int col=0; col < n; col++) {
            matrixRow += String.valueOf(a[row][col]) + " ";
         }
         matrix += matrixRow + "\n";
      }
      return matrix;
   }

 }

File: matrixMult.cpp
#include <iostream>
#include <chrono>
#include <ctime>
using namespace std;

//Matrix structure
struct Matrix {
	double values[5][5];
};

//Function Prototypes
Matrix multiply(Matrix &m1, Matrix &m2);
void toString(Matrix &m1);

/**
    This is the matrix multiplication method on nxn matrices.
    @param radius The radius of the circle.
    @return The volume of the sphere.
*/
Matrix multiply(Matrix &m1, Matrix &m2) {
	Matrix product;

	for (int row=0; row < 5; row++) {
		for (int col=0; col < 5; col++) {
			for(int i=0; i < 5; i++) {
				product.values[row][col] += (m1.values[row][i] * m2.values[i][col]);
			}
		}
	}
	return product;
}

/**
    This method turns the matrix into a printable value
	@param double a[][] A matrix.
	@return String 
*/
void toString(Matrix &m1) {
	for (int row=0; row < 5; row++) {
		for (int col=0; col < 5; col++) {
			cout << m1.values[row][col] << " ";
		}
		cout << "\n";
	}
	cout << "\n";
}

int main()
{
	Matrix arr1, arr2;

	//Matrix1
	arr1.values[0][0] = 1.0;
	arr1.values[0][1] = 2.0;
	arr1.values[0][2] = 3.0;
	arr1.values[0][3] = 4.0;
	arr1.values[0][4] = 5.0;

	arr1.values[1][0] = 6.0;
	arr1.values[1][1] = 7.0;
	arr1.values[1][2] = 8.0;
	arr1.values[1][3] = 9.0;
	arr1.values[1][4] = 10.0;

	arr1.values[2][0] = 11.0;
	arr1.values[2][1] = 12.0;
	arr1.values[2][2] = 13.0;
	arr1.values[2][3] = 14.0;
	arr1.values[2][4] = 15.0;

	arr1.values[3][0] = 16.0;
	arr1.values[3][1] = 17.0;
	arr1.values[3][2] = 18.0;
	arr1.values[3][3] = 19.0;
	arr1.values[3][4] = 20.0;

	arr1.values[4][0] = 21.0;
	arr1.values[4][1] = 22.0;
	arr1.values[4][2] = 23.0;
	arr1.values[4][3] = 24.0;
	arr1.values[4][4] = 25.0;

	//Matrix2
	arr2.values[0][0] = 1.0;
	arr2.values[0][1] = 2.0;
	arr2.values[0][2] = 3.0;
	arr2.values[0][3] = 4.0;
	arr2.values[0][4] = 5.0;

	arr2.values[1][0] = 6.0;
	arr2.values[1][1] = 7.0;
	arr2.values[1][2] = 8.0;
	arr2.values[1][3] = 9.0;
	arr2.values[1][4] = 10.0;

	arr2.values[2][0] = 11.0;
	arr2.values[2][1] = 12.0;
	arr2.values[2][2] = 13.0;
	arr2.values[2][3] = 14.0;
	arr2.values[2][4] = 15.0;

	arr2.values[3][0] = 16.0;
	arr2.values[3][1] = 17.0;
	arr2.values[3][2] = 18.0;
	arr2.values[3][3] = 19.0;
	arr2.values[3][4] = 20.0;

	arr2.values[4][0] = 21.0;
	arr2.values[4][1] = 22.0;
	arr2.values[4][2] = 23.0;
	arr2.values[4][3] = 24.0;
	arr2.values[4][4] = 25.0;

    //Matrix Multiplication 
    Matrix arr3 = multiply(arr1, arr2);
    toString(arr3);

    //Run-time Analysis
    std::chrono::steady_clock::time_point t1;
    std::chrono::steady_clock::time_point t2;
    std::chrono::duration<double> time_span;
    for(int i=0; i<10; i++) {
		t1 = std::chrono::steady_clock::now();
		Matrix arr3 = multiply(arr1, arr2);
		t2 = std::chrono::steady_clock::now();
		
		time_span += std::chrono::duration_cast<std::chrono::duration<double> >(t2 - t1);
		
	}
	std::chrono::duration<double> averageTime = time_span/10.0;
	averageTime *= 1000000000;	//Convert to nanoseconds

	std::cout << "C++ Matrix Mult Average time taken: " << averageTime.count() << " nanoseconds.";
	std::cout << std::endl;
}





